### ISSUE TITLE
EVENT_DISPATCH triggered twice!

### ISSUE THREAD

**Issue by gauthier** [zendframework/zendframework#7473]
While hunting a quite weird behavior related to some callback I connected to the MVC event named 'dispatch' (later referred as EVENT_DISPATCH), I found out that this event is triggered by `Zend\Mvc\Application::run()`. Makes sense. 

Then, the attached `DispatchListener` calls `dispatch()` on the controller. So far, so good...

But in the default `AbstractController::dispatch()`implementation, the same event EVENT_DISPATCH is triggered once again, despite we're still in the middle of the first triggering, so that the second one is *nested*. Makes no sense to me. 

Also note that the target of the first triggered event is the `Zend\Mvc\Application` while the target of the second one is the controller instance. This is means that depending on the priority provided when attaching a callback to an event, the callback can be related either to the Application triggering or the Controller one's (with any priority > 1).

I definitely think those two event should be split in two different events:

MvcEvent::EVENT_RUN (for instance) triggered by Application::run()
MvcEvent::EVENT_DISPATCH triggered by Controller::dispatch()

The typical execution flow would then be: 

```
BOOTSTRAP
ROUTE
RUN 
DISPATCH 
RENDER 
FINISH
```

Instead of

```
BOOTSTRAP
ROUTE
DISPATCH 
    -> DISPATCH 
RENDER 
FINISH
```

**Comment by adamlundrigan** [comment]
Your analysis is correct; ZendDeveloperTools reports the same for a stock ZendSkeletonApplication install:

![image](https://cloud.githubusercontent.com/assets/527329/7447681/d0c3ffea-f1df-11e4-8c22-cb95e9e9ce9d.png)

My assumption here is that the triggers are scoped, as attaching a listener to `MvcEvent::EVENT_DISPATCH` via the main Event Manager doesn't result in a double-trigger as you would expect if the event was being triggered in duplicate.  You can either attach to the Application's dispatch event or the Controller's dispatch event and your listener is scoped to the target you've chosen.  

Are you seeing any particular unwanted side-effect?  I agree that on the surface it's a bit counter-intuitive for two dispatch events to be triggered, but as far as I'm aware it doesn't cause any issues.

**Comment by Ocramius** [comment]
This is normal, tbh.

`dispatch` is triggered once on the application and once on the controller.

Issues may occur only if you don't set the correct event identifiers when attaching to the `SharedServiceManager` instance.

**Comment by gauthier** [comment]
@adamlundrigan yes there is a side-effect, which is major to me: depending on the priority you set when attaching a callback to `MvcEvent::EVENT_DISPATCH` you can't predict accurately to which triggering of this event you're callback will actually be attached, thus you can't know whether you'll get an instance of `Zend\Mvc\Application` or a controller. That said, you're right, the callback will run only once, but you don't know when (at which precise step of the workflow).

@Ocramius as said just above, it doesn't sound normal to me not being able to know precisely which event will run my callback. I don't understand what you mean by not setting the correct event identifier. Here, the event identifier is the same for both triggering.

This workflow is confuse, and I definitely don't understand the benefit of this nested event triggering. In my opinion, there is some mix-up between what does the application has to do at this step of the workflow (shunting the request to the appropriate controller) and what the controller has to do (actually dispatching the request by running the requested action)

**Comment by Martin-P** [comment]
With an indentifier you can tell the `EventManager` to only execute the event when it is called from a certain class. See the manual: http://framework.zend.com/manual/current/en/tutorials/tutorial.eventmanager.html

> thus you can't know whether you'll get an instance of `Zend\Mvc\Application` or a controller

That is only if you use a wildcard (`*`) as identifier. You can set a class as identifier (e.g. a controller) or you can use `$event->getTarget()` and determine if the target is an instance of `Zend\Mvc\Application` (or test for another class) and return early (or not).
```php
public function myCallbackMethod(MvcEvent $event)
{
    if ($event->getTarget() instanceof \Zend\Mvc\Application) {
        return;
    }

    // Code for targets which are not an instance of \Zend\Mvc\Application
}
```

**Comment by gauthier** [comment]
@Martin-P sure I can test the target from my callback but... I'm still not understanding why things have been done this way...

Regarding the event identifiers, firs, I was confused by what @Ocramius wrote:

> ...if you don't set the correct event identifiers when attaching...

I thought Marco talked about the name/key of the event itself, not the event emitter (by the way, one more confusion... `event identifiers` doesn't actually identify events, but event emitters :)), since you can't set events identifiers when attaching a callback!

Now that we come to that, it becomes even stranger, since I do not attach my callback to the `SharedEventManager`but to `EventManager` instance which is local to Application.

```
$eventManager->attach(MvcEvent::EVENT_DISPATCH, [$this, 'myCallback']);
```

Nevertheless, it is triggered from the Controller with a priority equal to or lower than 1, and from `Zend\Mvc\Application` with a priority greater than 1.

**Comment by GeeH** [comment]
This issue has been moved from the `zendframework` repository as part of the bug migration program as outlined here - http://framework.zend.com/blog/2016-04-11-issue-closures.html
New issue can be found at: https://github.com/zendframework/zend-mvc/issues/170
